End of round stats

Pspritechologist opened the pull request:
# Description
Adds a precedence for systems that want to add misc stats to the end of round screen, as well as a few such systems that add some interesting numbers!

# TODO
- [x] Mopping statistics
- [x] Top instrument player statistic
- [x] Amount bled statistics
- [x] Horns honked and Plushies squeaked statistics
- [x] Most time spent in cuffs statistic
- [x] Custom stats added through Admin commands
- [x] Bullets fired statistics
- [x] Enough statistics to be satisfying for now
- [x] Localize all of them


# Changelog

:cl:
- add: CentCom realized they weren't properly tracking you, and decided to update the telemetry terms of your contracts,. Expect to see some interesting stats at the end of the shift regarding things you may have, or may have not done well!

Pspritechologist committed:
Working Mopping stats and Music Played stats! FTLs to come

Pspritechologist committed:
Added total blood lost stat, total shots fired stat, and refined both mopping stats and instrument played stats. FTLs still TBD

Pspritechologist committed:
Minor tweak

Pspritechologist committed:
Several fixes post-update, cleaned up some lines, formatted things more nicer.

Pspritechologist committed:
Added stats for the player who spent the longest in cuffs, and for how many times sound was emitted from specifically tagged entities. New tags can be added extremely easily, current ones are BikeHorn and Plushie. ALL stats are now localized, and done well :))

Made it so cuff and instrument stats check for people currently cuffed or playing at the end of the round, as to not miss them.

Pspritechologist commented on Resources/Prototypes/SimpleStation14/tags.yml:
Alphabetical :)

Pspritechologist committed:
Removed an unneeded using, and clarified a comment. Should probably make a lot more comments...

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/BloodLost/BloodLostStatSystem.cs:
```diff
@@ -0,0 +1,46 @@
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<BloodLostStatEvent>(OnBloodLost);
```
you can probably just reference the system and set `totalBloodLost` to whatever in other places, would need the var to be public too

same with all the other stuff

Pspritechologist replied:
I feel doing these with events is better, as it's much more hands off with their respective systems. I can see how this one in particular, as well as shots fired could seem logical to do that way, but at that point it's simply about being consistent.

I find making an event for each one and allowing each system to simply log the information, send it off, and be done is cleaner.

I could, however, see the justification for a broader, more general system that manages *all* of the end round stats.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/EmitSound/EmitSoundStatSystem.cs:
```diff
@@ -0,0 +1,99 @@
+
+    // This Enum must match the exact tag you're searching for.
+    // Adding a new tag to this Enum, and ensuring the localisation is set will automatically add it to the end of round stats.
+
```
```suggestion
```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/EmitSound/EmitSoundStatSystem.cs:
```diff
@@ -0,0 +1,99 @@
+        Plushie
+    }
+
+    [Dependency] private readonly TagSystem _tag = default!;
```
to the top of the system

Pspritechologist replied:
I've seen precedence for doing it this way, global variables above dependencies. I can switch it.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/InstrumentPlayed/InstrumentPlayedStatSystem.cs:
```diff
@@ -0,0 +1,100 @@
+        public String? Username;
+    }
+
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
```
to the top of the system

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/ShotsFired/ShotsFiredStatSystem.cs:
```diff
@@ -0,0 +1,50 @@
+    private void OnShotFired(EntityUid _, GunComponent __, AmmoShotEvent args)
+    {
+        shotsFired++;
+    }
```
```suggestion
    }

```

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/ShotsFired/ShotsFiredStatSystem.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        var line = "\n[color=cadetblue]";
+
+        if (shotsFired < 25 && shotsFired != 0)
```
```suggestion
        if (shotsFired < 25)
```

Pspritechologist replied:
Huh? Why?

DEATHB4DEFEAT replied:
< 25 would cover the !0

Pspritechologist replied:
If it's less than 25 AND doesn't equal zero. So, if it's 1 through 24, or below 0.

Pspritechologist committed:
Apply suggestions from code review

Co-authored-by: DEATHB4DEFEAT <77995199+DEATHB4DEFEAT@users.noreply.github.com>

Pspritechologist committed:
Cleaned up several smaller things, and made everything use CCVars! :D

Pspritechologist committed:
Fixed broken FTL

Pspritechologist committed:
eof??
Added times slipped uhhhhh don't remember if it works yet. Probably doesn't.

Pspritechologist committed:
Added a command to allow Admins to add custom end of round stats!

Good for situations where you've given someone a custom objective, and want it to display.
Didn't bother localizing this one yet because command localizations are whack and we need to decide how to handle it.

Pspritechologist committed:
Cleaned up a couple unneeded usings.

Pspritechologist committed:
Slipping actually works now.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/Command/EORStatsListCommand.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    [Dependency] private readonly IEntityManager _entities = default!;
+    [Dependency] private readonly IEntitySystemManager _entitysys = default!;
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```
```suggestion
    [Dependency] private readonly IPrototypeManager _prototype = default!;
```

Pspritechologist replied:
This seems _highly_ unnecessary

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/Command/EORStatsListCommand.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        var _stats = IoCManager.Resolve<IEntitySystemManager>().GetEntitySystem<CommandStatSystem>();
+
+        if (args.Length != 0)
+        {
+            shell.WriteError("Invalid amount of arguments.");
+            return;
+        }
```
unneeded

Pspritechologist replied:
I like the commands being a little more strict. It's unneeded, but makes clear what's happening. Why are you passing additional arguments?

DEATHB4DEFEAT replied:
why does it need to be strict? arguments aren't used, no reason to complain about their validity

Pspritechologist replied:
It's to clarify usage (for instance, /kill might kill the user of the command only, but passing /kill <other username> should _not_ do that, just to avoid accidental misuse.). I realize that this command is super low risk, but I like following precedence.

DEATHB4DEFEAT commented on Content.Server/SimpleStation14/EndOfRoundStats/Command/EROStatsRemoveCommand.cs:
```diff
@@ -0,0 +1,74 @@
+
+        if (args.Length == 1)
+        {
+            var options = _stats.eorStats.Select(o => new CompletionOption
+            (
+                (_stats.eorStats.LastIndexOf((o.Item1, o.Item2)) + 1).ToString(), o.Item1
+            ));
```
wonk

Pspritechologist replied:
What about it?

DEATHB4DEFEAT replied:
just look funky

Pspritechologist replied:
wh
What about it? ;-;

Pspritechologist commented:
@DEATHB4DEFEAT Are tests working? Do I need to care about these fails?

DEATHB4DEFEAT commented:
> @DEATHB4DEFEAT Are tests working? Do I need to care about these fails?

it's unrelated

Pspritechologist committed:
General cleanup, very minor changes made.

Pspritechologist commented:
@DEATHB4DEFEAT This should be ready to merge, with your review.

Pspritechologist commented:
Still unsure about localizing the various commands, as there's no great formatting for it.

DEATHB4DEFEAT reviewed (approved):
lgtm
